solr/core/src/java/org/apache/solr/search/SortedIntDocSet.java

```diff
@@ -671,107 +679,141 @@ public DocSet union(DocSet other) {
     return new BitDocSet(newbits);
   }
 
-  @Override
-  public Filter getTopFilter() {
-    return new Filter() {
+  private int[] cachedOrdIdxMap;
```

Comment:
Ah, I had thought you meant to effectively build this "`ordIdxMap`" _in_ the ctor, as opposed to provided as an explicit arg to the `SortedIntDocSet` ctor! I'll look into what it would take to do the latter. 
The idea of building this during collection probably changes things, but I went the "cached" route initially (knowing that it could be moved into the ctor) to avoid doing extra work in the case of DocSets that may only used at the global level, and never asked to supply any `DocIdSetIterator` (e.g., in [JoinQuery.getDocSetEnumerate()](https://github.com/apache/solr/blob/7ada4032180b516548fc0263f42da6a7a917f92b/solr/core/src/java/org/apache/solr/search/JoinQuery.java#L454-L457)).

Reply:
Thanks for the review, @dsmiley!

I just pushed a handful of commits addressing all comments (I think) except for this one. I'll continue to mull this, but I'm curious how you're still feeling about it. My hesitance is based on the fact that accumulating this extra metadata during collection feels more invasive and potentially _less_ clean. Supplying as ctor arg also raises the issue of "merging" this metadata when a SortedIntDocSet is created as the result of an intersection or other set operation.

I'm still game to try, and maybe I'm making this out to be more complicated than it would really be; but at the moment all the code for this is in one place, and it feels relatively clean. With regard to concerns about caching ... this metadata is "cached", in a sense, whether it's lazily computed and cached (as here), precomputed in the ctor (using essentially the same logic), or supplied as a ctor arg (having been built during collection or some set operation). I'm confident that the current code is thread-safe, so perhaps the benefits of lazy eval outweigh the costs?

Reply:
Achieving thread-safety without 'volatile', 'synchronized', or some higher level things that tend to use them is tricky.  I see that this array ought to always have the same content.  If this were an `int` or other 4-byte primitive or a final/immutable object then I'd be comfortable.  But arrays... I'm not sure.  It'd be good to get another opinion.

I don't think computing this up-front always is a waste of time.  But I can see that asking the caller to do this is not as clean as keeping this matter internal, but moreover it's places other than DocSetBuilder that are harder to figure out.  So it's not important to me provided the caching works.